docs/source/internals/rest_api.rst

```diff
@@ -1,11 +1,272 @@
-.. todo::
+.. role:: python(code)
+   :language: python
 
-    .. _internal_architecture:rest-api:
+.. _internal_architecture:restapi:
 
-    ********
-    REST API
-    ********
+********
+REST API
+********
 
-    `#4043`_
+The AiiDA REST API is made of two main classes:
 
-.. _#4043: https://github.com/aiidateam/aiida-core/issues/4043
+    - ``App``, inheriting ``flask.Flask``. The latter represents any
+      Flask web app, including REST APIs.
+    - ``Api``, inheriting ``flask_restful.Api``. This represents the API itself.
+
+Once instantiated, both ``Api`` and ``App`` objects classes into, say, ``app`` and ``api``, these two objects have to be coupled by adding ``app`` as one of the attributes of ``api``.
+
+
+Extending the REST API
+======================
+
+In the following, we will go through a minimal example of creating an API that extends the AiiDA REST API by adding an endpoint ``/new-endpoint``.
+The endpoint will two HTTP methods:
+
+    - *GET*: retrieves the latest created Dict object and returns its ``id``,
+      ``ctime`` in ISO 8601 format, and ``attributes``.
+    - *POST*: creates a ``Dict`` object with placeholder attributes,
+      stores it, and returns its ``id``.
+
+In order to achieve this, we will need to:
+
+    - Create the ``flask_restful.Resource`` class that will be bound to the new endpoint.
+    - Extend the :py:class:`~aiida.restapi.api.AiidaApi` class in order to register the new endpoint.
+    - (Optional) Extend the :py:class:`~aiida.restapi.api.App` class for additional customization.
+
+Let's start by putting the following code into a  file ``api.py``:
+
+.. literalinclude:: includes/snippets/api.py
+
+We will now go through the previous code step by step.
+
+First things first: the imports.
+
+.. code-block:: python
+
+    from aiida.restapi.api import AiidaApi, App
+    from aiida.restapi.run_api import run_api
+    from flask_restful import Resource
+
+To start with, we import the base classes to be extended/employed: ``AiidaApi`` and ``App``.
+For simplicity, it is advisable to import the method ``run_api``, as it provides an interface to configure the Api, parse command-line arguments, and couple the two classes representing the Api and the App.
+However, you can refer to the documentation of `flask_restful <https://flask-restful.readthedocs.io/>`_ to configure and hook-up an Api through its built-in methods.
+
+Then we define a class representing the additional resource:
+
+.. code-block:: python
```

Comment:
> I see you created a separate file with the script: why do you have literal copies of it here instead of using includes of that file (or of parts of it)?

It was done this way in the original documentation (actually there, it was copied twice in the RST).
See https://github.com/aiidateam/aiida-core/pull/4492#issuecomment-712946515 for why I didn't change this.

> On the other hand, is this supposed to work if executed as is? Did you test if it still works this way?

Yes, for me it works

> I am getting a `OSError: [Errno 49] Can't assign requested address` when trying to follow the instructions (after `host`=>`hostname` fix), but maybe this is just me?

Not sure which fix you are referring to, but your error basically says that your port is occupied already - probably another REST API running.

Reply:
So, it turns out that probably nobody every actually tried the cli options suggested - the reason it fails is because it does not like 127.0.0.2 for a host. I'll fix it, thanks for checking!